src/spox/_attributes.py

```diff
@@ -184,38 +200,56 @@ def maybe(
     ) -> Optional[AttrIterableT]:
         return cls(tuple(value)) if value is not None else None
 
+    def _to_onnx_deref(self, key: str) -> AttributeProto:
+        return make_attribute(key, self.value, attr_type=self._attribute_proto_type_int)
+
 
 class AttrFloat32s(_AttrIterable[float]):
     _attribute_proto_type_int = AttributeProto.FLOATS
 
     def _to_onnx_deref(self, key: str) -> AttributeProto:
-        # ensure values are all floats
-        values = [float(v) for v in self.value]
-        return make_attribute(key, values, attr_type=AttributeProto.FLOATS)
+        try:
+            transformed = tuple(float(v) for v in self.value)
+        except ValueError as e:
+            raise AttributeTypeError("Attribute values don't seem to be floats.") from e
```

Comment:
> Also, the non-imperative language makes it sound like the real cause of the exception is unknown

It's more of a best effort for now since I'm not checking the exception's message.

> Why only catch ValueError, as in float("abc"), but not TypeError, as in float([])

👍 

> Why not use _get_pretty_type_exception here? Same applies to AttrStrings.

`_get_pretty_type_exception` is more specialized to the validation. In practice, we should indeed reach `_to_onnx_deref` from the `_validate` call first, but it's also called later on, so "Unable to instantiate `{type(self).__name__}` with value of type `{value_type}`." won't apply that well anymore.